### PR TITLE
[Component] Brightcove Video Player

### PR DESCRIPTION
**Pull request by jackharvey1**

### CODE REVIEW

File: packages/brightcove-video/__tests__/brightcove-video.android.test.js [hunk 1]
```diff
@@ -0,0 +1,11 @@
+import "react-native";
+import React from "react";
+import BrightcoveVideo from "../brightcove-video.android";
+import renderer from "react-test-renderer";
+
+it("renders correctly", () => {
+  const tree = renderer.create(<BrightcoveVideo />).toJSON();
+
+  expect(tree).toBeTruthy();
```
Comment (craigbilner) [inline packages/brightcove-video/__tests__/brightcove-video.android.test.js]: Can we change this to a snapshot test and split out the tests?

File: packages/brightcove-video/brightcove-video.web.js [hunk 2]
```diff
@@ -0,0 +1,57 @@
...
+
+let index = 0;
+
+class BrightcoveVideo extends Component {
+  componentDidMount() {
+    if (!BrightcoveVideo.hasLoadedScript) {
+      const s = document.createElement("script");
+      BrightcoveVideo.hasLoadedScript = true;
+
+      s.src = `//players.brightcove.net/${this.props
```
Comment (craigbilner) [inline packages/brightcove-video/brightcove-video.web.js]: Do we need an SRI?

File: packages/brightcove-video/brightcove-video.web.js [hunk 3]
```diff
@@ -0,0 +1,57 @@
...
+      return;
+    }
+    this.init();
+  }
+
+  init() {
+    if (window.bc) {
+      bc(document.getElementById(this.id));
+      videojs(this.id);
+    }
```
Comment (craigbilner) [inline packages/brightcove-video/brightcove-video.web.js]: Can we handle the else branch?

File: packages/brightcove-video/brightcove-video.web.js [hunk 4]
```diff
@@ -0,0 +1,57 @@
...
+      document.body.appendChild(s);
+      return;
+    }
+    this.init();
+  }
+
+  init() {
+    if (window.bc) {
+      bc(document.getElementById(this.id));
+      videojs(this.id);
```
Comment (craigbilner) [inline packages/brightcove-video/brightcove-video.web.js]: Does a truthy `window.bc` equate to a global `videojs`?

File: packages/brightcove-video/brightcove-video.stories.js [hunk 5]
```diff
@@ -0,0 +1,99 @@
...
+          }}
+        >
+          <Text style={{ color: "white" }}>click here to add a video</Text>
+        </TouchableOpacity>
+      </View>
+    );
+  }
+}
+
+storiesOf("BrightcoveVideo", module)
```
Comment (craigbilner) [inline packages/brightcove-video/brightcove-video.stories.js]: Could we add the error states here too?

File: packages/brightcove-video/brightcove-video.stories.js [hunk 6]
```diff
@@ -0,0 +1,99 @@
+import React, { Component } from "react";
+import { View, TouchableOpacity, Text } from "react-native";
+import { storiesOf } from "@storybook/react-native";
+import BrightcoveVideo from "./brightcove-video";
+
+const policyId =
+  "BCpkADawqM0NK0Rq8n6sEQyWykemrqeSmIQqqVt3XBrdpl8TYlvqN3hwKphBJRnkPgx6WAbozCW_VgTOBCNf1AQRh8KnmXSXfveQalRc5-pyNlSod5XzP99If2U";
+const accountId = "57838016001";
+const videoId = "4084164751001";
```
Comment (craigbilner) [inline packages/brightcove-video/brightcove-video.stories.js]: Could we see some other videos too "just in case" this is the only video that works?

File: packages/brightcove-video/brightcove-video.native.js [hunk 7]
```diff
@@ -0,0 +1,22 @@
+import React, { Component } from "react";
+import { requireNativeComponent } from "react-native";
+
+class BrightcoveVideo extends Component {
+  render() {
+    const height = this.props.height || "100%";
```
Comment (craigbilner) [inline packages/brightcove-video/brightcove-video.native.js]: are these `defaultProps`?

File: packages/brightcove-video/__tests__/brightcove-video.web.test.js [hunk 8]
```diff
@@ -0,0 +1,105 @@
...
+import renderer from "react-test-renderer";
+
+describe("brightcove-video web component", () => {
+  let appendChildMock, dummyElem;
+
+  beforeEach(() => {
+    appendChildMock = jest.fn();
+    dummyElem = {};
+
+    global.document = {
```
Comment (craigbilner) [inline packages/brightcove-video/__tests__/brightcove-video.web.test.js]: as a rule I think it's bad to mock the DOM like this, and integration tests of this nature should make use of `jsdom`. You can set the `testEnvironment` to `jsdom` which should give you this for for free and allow you test the behaviour rather than specific API calls

File: packages/brightcove-video/__tests__/brightcove-video.web.test.js [hunk 9]
```diff
@@ -0,0 +1,105 @@
...
+    window.bc = undefined;
+    window.videojs = undefined;
+    BrightcoveVideo.hasLoadedScript = false;
+    delete global.document;
+  });
+
+  it("renders correctly", () => {
+    const tree = renderer.create(<BrightcoveVideo />).toJSON();
+
+    expect(tree).toBeTruthy();
```
Comment (craigbilner) [inline packages/brightcove-video/__tests__/brightcove-video.web.test.js]: change to snapshot test

File: packages/brightcove-video/__tests__/brightcove-video.web.test.js [hunk 10]
```diff
@@ -0,0 +1,105 @@
...
+  it("passes accountId & videoId to video correctly", () => {
+    const tree = renderer
+      .create(<BrightcoveVideo accountId="[ACCOUNT_ID]" videoId="[VIDEO_ID]" />)
+      .toJSON();
+
+    expect(tree.children[0].props["data-video-id"]).toBe("[VIDEO_ID]");
+    expect(tree.children[0].props["data-account"]).toBe("[ACCOUNT_ID]");
+  });
+
+  it("does not attempt to initialise the brightcove player before the script has loaded", () => {
```
Comment (craigbilner) [inline packages/brightcove-video/__tests__/brightcove-video.web.test.js]: the test name look good but the test code doesn't look like it matches? One problem might be the video throwing and it implying it was a script issue when it may not be?

**Comment by craigbilner** [comment]
we need a LICENCE too :-)

**Comment by jackharvey1** [comment]
@craigbilner of course, but separate PR please :)

**Comment by trevorah** [comment]
just had a tricky thing with @RussellToon when trying to use this package:

`import BrightcovePlayer from '@newsint/brightcove-player';` does not work.
`import BrightcovePlayer from '@newsint/brightcove-player/brightcove-player';` _does_ work.

Looks like the packager is having problems with the package.json `main`.

Sidenote for any future testers of pre-published components: npm messes up file dependencies, yarn doesn't. That'll save you hours of pain.

File: packages/brightcove-video/brightcove-video.web.js (hunk 2 above)
Comment (dickie81) [inline packages/brightcove-video/brightcove-video.web.js]: What happens if Brightcove decide that they want to update their player? We would need to update the SRI in sync with their release? I think this would be possible only if we know that they will never update their player code.

File: packages/brightcove-video/brightcove-video.native.js (hunk 7 above)
Comment (aronshamash) [inline packages/brightcove-video/brightcove-video.native.js]: yes we should use that

File: packages/brightcove-video/brightcove-video.web.js (hunk 2 above)
Comment (aronshamash) [inline packages/brightcove-video/brightcove-video.web.js]: I am chasing Brightcove to support SRI, but I'd be surprised if they would be able to implement quickly enough for this PR.

File: packages/brightcove-video/brightcove-video.stories.js (hunk 6 above)
Comment (dickie81) [inline packages/brightcove-video/brightcove-video.stories.js]: I have added another video - it was the only other brightcove test video I could find on the same account. It seems to have problems if you try to load videos from another account. I guess if we want to be able to support multiple accounts on the same page we would need to load another script as the script URL includes the account id. Should we add this to issues or try to fix? I don't think we will need to load videos from multiple accounts while using on the Times web site.

File: packages/brightcove-video/__tests__/brightcove-video.web.test.js (hunk 10 above)
Comment (dickie81) [inline packages/brightcove-video/__tests__/brightcove-video.web.test.js]: Could we add Sinon or similar for easier spies / stubs? It would make testing stuff like this easier & cleaner.

File: packages/brightcove-video/brightcove-video.stories.js (hunk 6 above)
Comment (craigbilner) [inline packages/brightcove-video/brightcove-video.stories.js]: Raise an issue 👍

File: .storybook/webpack.config.js [hunk 11]
```diff
@@ -10,6 +10,6 @@ module.exports = {
     // If you're working on a multi-platform React Native app, web-specific
```
Comment (dickie81) [inline .storybook/webpack.config.js]: Still a useful hint?

File: packages/brightcove-video/brightcove-video.stories.js (hunk 6 above)
Comment (dickie81) [inline packages/brightcove-video/brightcove-video.stories.js]: raised issue here: #30

File: packages/brightcove-video/__tests__/brightcove-video.web.test.js (hunk 10 above)
Comment (dickie81) [inline packages/brightcove-video/__tests__/brightcove-video.web.test.js]: found jest spyOn and improved test

File: packages/brightcove-video/__tests__/__snapshots__/brightcove-video.ios.test.js.snap [hunk 12]
```diff
@@ -0,0 +1,16 @@
+// Jest Snapshot v1, https://goo.gl/fbAQLP
```
Comment (jackharvey1) [inline packages/brightcove-video/__tests__/__snapshots__/brightcove-video.ios.test.js.snap]: Do we need these comments?

File: .storybook/webpack.config.js (hunk 11 above)
Comment (jackharvey1) [inline .storybook/webpack.config.js]: I'd rather not have them in personally

File: packages/brightcove-video/android/rntbrightcovevideo/src/test/java/uk/co/news/rntbrightcovevideo/ExampleUnitTest.java [hunk 13]
```diff
@@ -0,0 +1,17 @@
...
+ * Example local unit test, which will execute on the development machine (host).
+ *
+ * @see <a href="http://d.android.com/tools/testing">Testing documentation</a>
+ */
+public class ExampleUnitTest {
+    @Test
+    public void addition_isCorrect() throws Exception {
+        assertEquals(4, 2 + 2);
+    }
+}
```
Comment (jackharvey1) [inline packages/brightcove-video/android/rntbrightcovevideo/src/test/java/uk/co/news/rntbrightcovevideo/ExampleUnitTest.java]: Do we still want to keep this?

File: packages/brightcove-video/brightcove-video.web.js [hunk 14]
```diff
@@ -0,0 +1,192 @@
...
+
+  createScript() {
+    const s = document.createElement("script");
+    s.src = BrightcoveVideo.getScriptUrl(this.props.accountId);
+
+    return s;
+  }
+
+  appendScript(s) {
+    document.body.appendChild(s);
```
Comment (jackharvey1) [inline packages/brightcove-video/brightcove-video.web.js]: Do we want this on the body? Why not the head?

File: packages/brightcove-video/brightcove-video.web.js [hunk 15]
```diff
@@ -0,0 +1,192 @@
...
+      );
+
+      return (
+        <View
+          style={{
+            width: this.props.width,
+            height: this.props.height,
+            backgroundColor: "red"
+          }}
+        >
```
Comment (jackharvey1) [inline packages/brightcove-video/brightcove-video.web.js]: God dammit prettify

File: .storybook/webpack.config.js (hunk 11 above)
Comment (jackharvey1) [inline .storybook/webpack.config.js]: I'd rather not have them in personally

File: packages/brightcove-video/ios/README.md [hunk 16]
```diff
@@ -0,0 +1,7 @@
+# iOS Setup
+
+To install the Brightcove SDK, follow Brightcove's instructions found [here](https://github.com/brightcove/brightcove-player-sdk-ios#manual). We recommend using `v5.3.3`, however any recent version should be fine.
+
+Navigate to `node_modules/@newsint/brightcove-video/ios/`, copying the two pairs of `.h` and `.m` files in to your project.
+
+When you are presented with the dialogue, _do not_ copy the items. You may choose to create a group or not depending on your preferences.
```
Comment (trevorah) [inline packages/brightcove-video/ios/README.md]: You would probably make everything easier by defining a pod spec with a dependency on https://cocoapods.org/pods/Brightcove-Player-SDK.

File: packages/brightcove-video/brightcove-video.web.js [hunk 17]
```diff
@@ -0,0 +1,192 @@
...
+          height={this.props.height}
+          data-embed="default"
+          data-video-id={this.props.videoId}
+          data-account={this.props.accountId}
+          data-player="default"
+          data-application-id
+          className="video-js"
+          controls
+        />
+      </div>
```
Comment (trevorah) [inline packages/brightcove-video/brightcove-video.web.js]: Why the wrapping div?

Comment (dickie81) [inline packages/brightcove-video/brightcove-video.web.js]: Brightcove SDK does not play nicely unless there is a wrapping div

File: packages/brightcove-video/ios/README.md (hunk 16 above)
Comment (dickie81) [inline packages/brightcove-video/ios/README.md]: Can this be part of another PR? I think Russ wants to see how he can improve the whole process probably by turning it into a library.

File: packages/brightcove-video/__tests__/__snapshots__/brightcove-video.ios.test.js.snap (hunk 12 above)
Comment (dickie81) [inline packages/brightcove-video/__tests__/__snapshots__/brightcove-video.ios.test.js.snap]: Auto generated - suggest not deleting as they will just be created again.

File: packages/brightcove-video/android/rntbrightcovevideo/src/test/java/uk/co/news/rntbrightcovevideo/ExampleUnitTest.java (hunk 13 above)
Comment (dickie81) [inline packages/brightcove-video/android/rntbrightcovevideo/src/test/java/uk/co/news/rntbrightcovevideo/ExampleUnitTest.java]: I think it is a useful pointer & reminder that we need to write tests - difficult thing is running them as part of CI / build.

File: packages/brightcove-video/brightcove-video.web.js (hunk 14 above)
Comment (dickie81) [inline packages/brightcove-video/brightcove-video.web.js]: Maybe?

File: packages/brightcove-video/brightcove-video.stories.js (hunk 5 above)
Comment (dickie81) [inline packages/brightcove-video/brightcove-video.stories.js]: Done

**Comment by dickie81** [comment]
Can we get a re-review & an idea of what if anything is blocking the merge of this PR?

File: packages/brightcove-video/package.json [hunk 18]
```diff
@@ -0,0 +1,56 @@
+{
+  "name": "@newsint/brightcove-video",
```
Comment (jackharvey1) [inline packages/brightcove-video/package.json]: should be `@timescomponents`

File: packages/brightcove-video/brightcove-video.web.js [hunk 19]
```diff
@@ -0,0 +1,192 @@
...
+    );
+  }
+}
+
+BrightcoveVideo.globalErrors = [];
+
+BrightcoveVideo.defaultProps = {
+  width: 320,
+  height: 180
```
Comment (trevorah) [inline packages/brightcove-video/brightcove-video.web.js]: Default props are different from native. Why would anyone use the defaults if they end up with different results between web and native?

Make them the same.

File: packages/brightcove-video/brightcove-video.web.js [hunk 20]
```diff
@@ -0,0 +1,192 @@
...
+      this.props.onError(err);
+    }
+  }
+
+  onError(player) {
+    this.emitError(player.error());
+  }
+
+  onPlay(player) {
+    this.setState({
```
Comment (craigbilner) [inline packages/brightcove-video/brightcove-video.web.js]: because these are all asynchronous, `setState` can be called on an unmounted component. You'll need to effectively cancel this events when the component unmounts. You also need to unsubscribe on unmount?

File: packages/brightcove-video/brightcove-video.stories.js [hunk 21]
```diff
@@ -0,0 +1,148 @@
...
+      <BrightcoveVideo
+        width={320}
+        height={200}
+        policyId={policyId}
+        videoId="1532562858001"
+        accountId={accountId}
+      />
+    </View>
+  )
+  .add("Bad video id", () =>
```
Comment (craigbilner) [inline packages/brightcove-video/brightcove-video.stories.js]: can you get a steering from Claire on what this state should be, can be fixed later but a red box saying "unknown" isn't ideal

File: packages/brightcove-video/brightcove-video.web.js [hunk 22]
```diff
@@ -0,0 +1,192 @@
...
+            backgroundColor: "red"
+          }}
+        >
+          <ul>{errorItems}</ul>
+        </View>
+      );
+    }
+
+    return (
+      <div>
```
Comment (trevorah) [inline packages/brightcove-video/brightcove-video.web.js]: Add a comment like "Wrapping div required as brightcove adds siblings to the video tag"

File: packages/brightcove-video/brightcove-video.web.js [hunk 23]
```diff
@@ -0,0 +1,192 @@
...
+  init() {
+    if (window.bc && window.videojs) {
+      this.initVideo(this.state.id);
+    } else {
+      BrightcoveVideo.players.push(this);
+    }
+  }
+
+  static getScriptUrl(accountId) {
+    return `//players.brightcove.net/${accountId}/default_default/index.min.js`;
```
Comment (craigbilner) [inline packages/brightcove-video/brightcove-video.web.js]: can we "https all the things"?

File: packages/brightcove-video/brightcove-video.stories.js [hunk 24]
```diff
@@ -0,0 +1,148 @@
...
+  )
+  .add("Bad account id", () =>
+    <BrightcoveVideo
+      width={800}
+      height={600}
+      policyId={policyId}
+      videoId={videoId}
+      accountId="x"
+      onError={firstArgJSONAction("error")}
+    />
```
Comment (trevorah) [inline packages/brightcove-video/brightcove-video.stories.js]: "Bad video id" and "Bad account id" on native show a video with clickable buttons that do nothing. This is different from web, which shows a red box with user unfriendly text. What should it show?

File: ios/BrightcovePlayerSDK.framework/Headers/NSDictionary+BCOVURLSupport.h [hunk 25]
```diff
@@ -0,0 +1,26 @@
...
+/**
+ * Constructs a string of key/value pairs that have been UTF-8 encoded for a
+ * URL.
+ *
+ * @return The URL encoded string form of this dictionary.
+ */
+- (NSString *)bcov_UTF8EncodedRequestParameterString;
+
+@end
+
```
Comment (jackharvey1) [inline ios/BrightcovePlayerSDK.framework/Headers/NSDictionary+BCOVURLSupport.h]: Yes we're going to lock down this PR feature-wise because it needs to be merged asap. Perhaps raise an issue?

File: packages/brightcove-video/brightcove-video.web.js [hunk 26]
```diff
@@ -0,0 +1,192 @@
...
+    document.body.appendChild(s);
+  }
+
+  static handlePlayerReady(context) {
+    this.on("play", context.onPlay.bind(context, this));
+    this.on("pause", context.onPause.bind(context, this));
+    this.on("seeked", context.onSeeked.bind(context, this));
+  }
+
+  emitState() {
```
Comment (dickie81) [inline packages/brightcove-video/brightcove-video.web.js]: can we leave this for a further PR when we have decided all events we require?

File: packages/brightcove-video/brightcove-video.stories.js (hunk 21 above)
Comment (dickie81) [inline packages/brightcove-video/brightcove-video.stories.js]: Can we leave this for a further PR when we have designs otherwise we are likely to be blocked?

File: packages/brightcove-video/brightcove-video.stories.js [hunk 27]
```diff
@@ -0,0 +1,148 @@
+import React, { Component } from "react";
+import { View, TouchableOpacity, Text } from "react-native";
+import { storiesOf } from "@storybook/react-native";
+import { action, decorateAction } from "@storybook/addon-actions";
```
Comment (dickie81) [inline packages/brightcove-video/brightcove-video.stories.js]: sure - will be great when we have eslint

File: packages/brightcove-video/brightcove-video.web.js [hunk 28]
```diff
@@ -0,0 +1,192 @@
...
+  render() {
+    if (this.state.errors.length) {
+      const errorItems = this.state.errors.map((error, i) =>
+        <li key={i} style={{ color: "white" }}>
+          {error.code} - {error.message}
+        </li>
+      );
+
+      return (
+        <View
```
Comment (aronshamash) [inline packages/brightcove-video/brightcove-video.web.js]: why do we need to stick to web elements in web files?

File: packages/brightcove-video/brightcove-video.native.js [hunk 29]
```diff
@@ -0,0 +1,42 @@
...
+    }
+  }
+}
+
+BrightcoveVideo.defaultProps = {
+  width: "100%",
+  height: "100%"
+};
+
+const RNTBrightcove = requireNativeComponent("RNTBrightcove", null);
```
Comment (dickie81) [inline packages/brightcove-video/brightcove-video.native.js]: You can of course.

This was taken from the example structure given in the examples. Would you prefer we did not follow this structure?

https://facebook.github.io/react-native/docs/native-components-ios.html

File: packages/brightcove-video/brightcove-video.native.js [hunk 30]
```diff
@@ -0,0 +1,42 @@
...
+  _onChange(event) {
+    if (this.props.onChange) {
+      this.props.onChange(event.nativeEvent);
+    }
+  }
+}
+
+BrightcoveVideo.defaultProps = {
+  width: "100%",
+  height: "100%"
```
Comment (jackharvey1) [inline packages/brightcove-video/brightcove-video.native.js]: React web doesn't iirc.

File: packages/brightcove-video/brightcove-video.stories.js (hunk 24 above)
Comment (dickie81) [inline packages/brightcove-video/brightcove-video.stories.js]: we have no design yet - do we want to block this until we do? as we have error events this could also be handled by any wrapping component so any design could be implemented not something Times specific (I don't think having Times specific errors is a good idea either)

File: packages/brightcove-video/brightcove-video.native.js [hunk 31]
```diff
@@ -0,0 +1,42 @@
...
+};
+
+const RNTBrightcove = requireNativeComponent("RNTBrightcove", null);
+
+export default BrightcoveVideo;
```
Comment (dickie81) [inline packages/brightcove-video/brightcove-video.native.js]: Craig advised us not to use propTypes as we will be using flow. Craig can you confirm this or have we changed direction?

File: packages/brightcove-video/brightcove-video.stories.js (hunk 24 above)
Comment (trevorah) [inline packages/brightcove-video/brightcove-video.stories.js]: Just make it the same big red square for both.

File: packages/brightcove-video/brightcove-video.native.js (hunk 29 above)
Comment (trevorah) [inline packages/brightcove-video/brightcove-video.native.js]: Ah, ok. I guess you'll need to call it like:
```js
const RNTBrightcove = requireNativeComponent("RNTBrightcove", BrightcoveVideo);
```

File: ios/BrightcovePlayerSDK.framework/Headers/NSDictionary+BCOVURLSupport.h (hunk 25 above)
Comment (trevorah) [inline ios/BrightcovePlayerSDK.framework/Headers/NSDictionary+BCOVURLSupport.h]: I can take a quick look, once its merged

File: packages/brightcove-video/brightcove-video.web.js [hunk 32]
```diff
@@ -0,0 +1,193 @@
+import React, { Component } from "react";
+import { View, Text } from "react-native";
+
+let index = 0;
+
+class BrightcoveVideo extends Component {
+  constructor(props) {
+    super(props);
+
+    index++;
```
Comment (trevorah) [inline packages/brightcove-video/brightcove-video.web.js]: please use a uuid rather than hoping that no other element on the page chooses "1" as an id.

Comment (dickie81) [inline packages/brightcove-video/brightcove-video.web.js]: its not quite that bad: `id: \`${props.videoId}-${props.accountId}-${index}\``

do we still need to change this?

File: packages/brightcove-video/brightcove-video.web.js (hunk 26 above)
Comment (jackharvey1) [inline packages/brightcove-video/brightcove-video.web.js]: #40

File: packages/brightcove-video/brightcove-video.web.js (hunk 19 above)
Comment (dickie81) [inline packages/brightcove-video/brightcove-video.web.js]: Done
